syntax/grammar.mjs

```diff
@@ -0,0 +1,496 @@
+        MessageAttributeExpression,
+        TermVariantExpression,
+        Identifier.map(into(FTL.MessageReference)),
+        TermIdentifier.map(into(FTL.MessageReference)),
+        ExternalArgument,
+        Placeable));
+
+/* ------------------ */
+/* Inline Expressions */
+let StringExpression = defer(() =>
+    quoted_text.map(into(FTL.StringExpression)));
+
+let NumberExpression = defer(() =>
+    number.map(into(FTL.NumberExpression)));
+
+let ExternalArgument = defer(() =>
+    ExternalIdentifier.map(into(FTL.ExternalArgument)));
+
+let CallExpression = defer(() =>
+    sequence(
+        Function.as("callee"),
+        char("("),
+        maybe(argument_list).as("args"),
+        maybe(inline_space),
+        char(")"))
+    .map(to_object)
+    .map(into(FTL.CallExpression)));
+
+let argument_list = defer(() =>
+    sequence(
+        Argument.as(),
+        repeat(
+            sequence(
+                char(","),
+                Argument.as())),
+        maybe(char(",")))
+    .map(flatten(2))
+    .map(to_array));
+
+let Argument = defer(() =>
+    between(
+        maybe(inline_space),
+        either(
+            NamedArgument,
+            InlineExpression)));
+
+let NamedArgument = defer(() =>
+    sequence(
+        Identifier.as("name"),
+        between(
+            maybe(inline_space),
+            char(":")),
+        either(
+            StringExpression,
+            NumberExpression).as("value"))
+    .map(to_object)
+    .map(into(FTL.NamedArgument)));
+
+let MessageAttributeExpression = defer(() =>
+    sequence(
+        Identifier.as("id"),
+        char("."),
+        Identifier.as("name"))
+    .map(to_object)
+    .map(into(FTL.AttributeExpression)));
+
+let TermAttributeExpression = defer(() =>
+    sequence(
+        TermIdentifier.as("id"),
+        char("."),
+        Identifier.as("name"))
+    .map(to_object)
+    .map(into(FTL.AttributeExpression)));
+
+let TermVariantExpression = defer(() =>
+    sequence(
+        TermIdentifier.as("id"),
+        char("["),
+        between(
+            maybe(inline_space),
+            VariantKey.as("key")),
+        char("]"))
+    .map(to_object)
+    .map(into(FTL.VariantExpression)));
+
+/* ----------------- */
+/* Block Expressions */
+let SelectExpression = defer(() =>
+    sequence(
+        SelectorExpression.as("selector"),
+        between(
+            maybe(inline_space),
+            string("->")),
+        variant_list.as("variants"),
+        break_indent)
+    .map(to_object)
+    .map(into(FTL.SelectExpression)));
+
+let VariantList = defer(() =>
+    sequence(
+        always(null).as("selector"),
+        variant_list.as("variants"),
+        break_indent)
+    .map(to_object)
+    .map(into(FTL.SelectExpression)));
+
+let variant_list = defer(() =>
+    sequence(
+        repeat(Variant),
+        DefaultVariant,
+        repeat(Variant))
+    .map(flatten(1)));
+
+let VariantName = defer(() =>
```

Comment:
`VariantName` and `VariantKey` threw me off here coming before `Variant` and `DefaultVariant`. Want to move them down?